src/Psalm/Internal/Analyzer/ClassAnalyzer.php

```diff
@@ -2627,5 +2635,119 @@ private function checkEnum(PhpParser\Node\Stmt\Enum_ $enum): void
                 }
             }
         }
+
+        $case_values = [];
+        $case_names = [];
+
+        foreach ($enum->stmts as $stmt) {
+            if ($stmt instanceof PhpParser\Node\Stmt\EnumCase) {
+                if (isset($case_names[$stmt->name->name])) {
+                    if (IssueBuffer::accepts(
+                        new DuplicateEnumCase(
+                            'Enum case names should be unique',
+                            new CodeLocation($this, $stmt),
+                            $enum->name->name
+                        )
+                    )) {
+                        continue;
+                    }
+                }
+
+                $case_names[$stmt->name->name] = true;
+
+                if (null !== $stmt->expr && null === $enum->scalarType) {
+                    if (IssueBuffer::accepts(
+                        new InvalidEnumCaseValue(
+                            'Case of a non-backed enum should not have a value',
+                            new CodeLocation($this, $stmt),
+                            $enum->name->name
+                        )
+                    )) {
+                        continue;
+                    }
+                } elseif (null === $stmt->expr && null !== $enum->scalarType) {
+                    if (IssueBuffer::accepts(
+                        new InvalidEnumCaseValue(
+                            'Case of a backed enum should have a value',
+                            new CodeLocation($this, $stmt),
+                            $enum->name->name
+                        )
+                    )) {
+                        continue;
+                    }
+                } elseif (null !== $enum->scalarType && null !== $stmt->expr) {
+                    if (ExpressionAnalyzer::analyze(
+                        new StatementsAnalyzer(
+                            $this,
+                            $type_provider = new \Psalm\Internal\Provider\NodeDataProvider()
```

Comment:
Can you explain why you had to do that? You couldn't fetch the existing one? Or was there not one yet in the execution?

Reply:
Because it was 2 AM, mostly =)

I probably didn't have to. There should be already an instance of enum storage populated that has `$enum_cases` property, and if that doesn't support cases like https://3v4l.org/YV9Xm/rfc#vgit.master it should be probably made to, using `SimpleTypeInferrer`.